clang/lib/Headers/__clang_hip_math.h

```diff
@@ -1306,14 +1306,50 @@ float min(float __x, float __y) { return __builtin_fminf(__x, __y); }
 __DEVICE__
 double min(double __x, double __y) { return __builtin_fmin(__x, __y); }
 
+// Define host min/max functions.
+
 #if !defined(__HIPCC_RTC__) && !defined(__OPENMP_AMDGCN__)
-__host__ inline static int min(int __arg1, int __arg2) {
-  return __arg1 < __arg2 ? __arg1 : __arg2;
-}
 
-__host__ inline static int max(int __arg1, int __arg2) {
-  return __arg1 > __arg2 ? __arg1 : __arg2;
-}
+#pragma push_macro("DEFINE_MIN_MAX_FUNCTIONS")
+#define DEFINE_MIN_MAX_FUNCTIONS(type1, type2) \
+static inline auto min(const type1 __a, const type2 __b) \
+  -> typename std::remove_reference<decltype(__a < __b ? __a : __b)>::type { \
+  return (__a < __b) ? __a : __b; \
+} \
+static inline auto max(const type1 __a, const type2 __b) \
+  -> typename std::remove_reference<decltype(__a > __b ? __a : __b)>::type { \
+  return (__a > __b) ? __a : __b; \
+}
+
+// Define min and max functions for same type comparisons
+DEFINE_MIN_MAX_FUNCTIONS(int, int)
```

Comment:
Could we not do something like this w/ the appropriate static assertion? Or is there an important restriction on the specific types for this function.
```c
template <typename T, typename U>
static inline auto min(const T &__a, const U &__b) {
  return (__a < __b) ? __a : __b;
}
```

Reply:
it will return reference of `__a` or `__b`. that won't work since it is reference to stack var

Reply:
sorry I was wrong about the reference.

The issue with the template approach is that it has subtle differences regarding overloading resolution compared with CUDA, which could lead to incompatibility.

Reply:
Could we not do stuff like `static_assert` where we check if both types are the same modulo `unsigned` or `const`? I'm assuming that would be similar. I'd just prefer to avoid macro magic if possible, but it we really need it for compatibility reasons we can use it.

Reply:
it is not just limiting types. template function has lower precedence than functions in overloading resolution. If users using std::min we will get different behavior than CUDA